p2p/enr/enr.go

```diff
@@ -0,0 +1,396 @@
+	e.dirty = true
+	e.records = append(e.records, record{k, v})
+}
+
+func (e *ENR) Encode() ([]byte, error) {
+	if e.dirty {
+		return nil, errors.New("record is not signed")
+	}
+	return e.raw, nil
+}
+
+func (e *ENR) NodeAddress() ([]byte, error) {
+	pk, err := e.GetSecp256k1()
+	if err != nil {
+		return nil, err
+	}
+
+	digest := crypto.Keccak256Hash(pk)
+
+	return digest.Bytes(), nil
+}
+
+func (e *ENR) Decode(data []byte) error {
+	if len(data) > RECORD_MAX_SIZE {
+		return errors.New("record is too big")
+	}
+
+	s := rlp.NewStream(bytes.NewReader(data), RECORD_MAX_SIZE)
+
+	signature, err := s.Bytes()
+	if err != nil {
+		return err
+	}
+
+	// consume the list prefix
+	_, err = s.List()
+	if err != nil {
+		return err
+	}
+
+	seq, err := s.Uint()
+	if err != nil {
+		return err
+	}
+
+	var records []record
+
+	// read key/value pairs until we reach rlp.EOL
+	for _, _, err = s.Kind(); err == nil; _, _, err = s.Kind() {
+		key, err2 := s.Bytes()
+		if err2 != nil {
+			return err2
+		}
+
+		value, err2 := s.Bytes()
+		if err2 != nil {
+			return err2
+		}
+
+		records = append(records, record{k: key, v: value})
+	}
+
+	if err != rlp.EOL {
+		return err
+	}
+
+	e.signature = signature
+	e.raw = data
+	e.seq = uint32(seq)
+	e.records = records
+
+	err = e.verifySignature()
+	if err != nil {
+		return err
+	}
+
+	e.dirty = false
+
+	return nil
+}
+
+func (e *ENR) Sign(privkey *ecdsa.PrivateKey) error {
+	e.seq = e.seq + 1
+
+	e.SetID(ID_SECP256k1_KECCAK)
+
+	pk := (*btcec.PublicKey)(&privkey.PublicKey).SerializeCompressed()
+	e.SetSecp256k1(pk)
+
+	var err error
+	e.signature, e.raw, err = e.SignAndEncode(privkey)
+	if err != nil {
+		return err
+	}
+
+	// mark record ready for encoding
+	e.dirty = false
+
+	return nil
+}
+
+func (e *ENR) SignAndEncode(privkey *ecdsa.PrivateKey) ([]byte, []byte, error) {
+	content, err := e.SerialisedContent()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	digest := crypto.Keccak256Hash(content)
+
+	signature, err := crypto.Sign(digest.Bytes(), privkey)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	blob, err := rlp.EncodeToBytes(signature)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	raw := append(blob, content...)
+
+	return signature, raw, nil
+}
+
+func (e *ENR) SerialisedContent() ([]byte, error) {
+	var buffer bytes.Buffer
+
+	blob, err := rlp.EncodeToBytes(e.seq)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = buffer.Write(blob)
+	if err != nil {
+		return nil, err
+	}
+
+	sort.Slice(e.records, func(i, j int) bool {
+		return bytes.Compare(e.records[i].k, e.records[j].k) < 0
+	})
+
+	for _, r := range e.records {
+		kk, err := rlp.EncodeToBytes(r.k)
+		if err != nil {
+			return nil, err
+		}
+
+		_, err = buffer.Write(kk)
+		if err != nil {
+			return nil, err
+		}
+
+		vv, err := rlp.EncodeToBytes(r.v)
+		if err != nil {
+			return nil, err
+		}
+
+		_, err = buffer.Write(vv)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return wrapList(buffer.Bytes()), nil
```

Comment:
You can avoid most of the RLP hackery by using higher-level types instead. To encode a record, put elements into `[]interface{}` then call 
`rlp.EncodeToBytes`. 

```go
list := []interface{}{nil, e.seq}
// add k/v elements
for _, r := range e.records {
    list = append(list, k, v)
}
// sign the tail of the list
sigcontent, err := rlp.EncodeToBytes(list[1:])
...
// add signature to front
list[0] = sig
record, err := rlp.EncodeToBytes(list)
```